src/main/java/com/hierynomus/smbj/transport/tcp/AsyncPacketReader.java

```diff
@@ -0,0 +1,190 @@
+/*
+ * Copyright (C)2016 - SMBJ Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hierynomus.smbj.transport.tcp;
+
+import java.io.EOFException;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.nio.channels.AsynchronousCloseException;
+import java.nio.channels.AsynchronousSocketChannel;
+import java.nio.channels.CompletionHandler;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.hierynomus.protocol.Packet;
+import com.hierynomus.protocol.commons.buffer.Buffer.BufferException;
+import com.hierynomus.smbj.transport.PacketFactory;
+import com.hierynomus.smbj.transport.PacketReader;
+import com.hierynomus.smbj.transport.PacketReceiver;
+import com.hierynomus.smbj.transport.TransportException;
+
+public class AsyncPacketReader<P extends Packet<P, ?>> {
+    private static final Logger logger = LoggerFactory.getLogger(PacketReader.class);
+
+    private final PacketFactory<P> packetFactory;
+    private PacketReceiver<P> handler;
+    private final AsynchronousSocketChannel channel;
+    private String remoteHost;
+    private int soTimeout = 0;
+
+    private AtomicBoolean stopped = new AtomicBoolean(false);
+
+    private static final int READ_BUFFER_CAPACITY = 9000; // Size of a Jumbo frame
+    private static final int HEADER_SIZE = 4;
+
+    private final ByteBuffer readBuffer;
+    private int bytesWaitingToBeProcessed = 0;
+
+    private static final int NO_PACKET_LENGTH = -1;
+
+    private int currentPacketLength = NO_PACKET_LENGTH;
```

Comment:
Indeed, that is possible. However I think you should encapsulate this logic instead of flattening it out in the PacketReader.